### PR TITLE
refactor(ir/api): introduce window frame operation and revamp the window API

### LINKED ISSUE #1834: feat: reconsider ibis's window API
I'd like to propose a new and more compact API for window frame specification that is heavily inspired by the SQLAlchemy way of spelling these:

```python
t = ibis.table(...)

# SELECT AVG(col) OVER (ROWS BETWEEN 1 PRECEDING AND CURRENT ROW) FROM t
t.col.mean().over(rows=(-1, 0))

# SELECT AVG(col) OVER (ROWS BETWEEN 1 PRECEDING AND CURRENT ROW) FROM t
t.col.mean().over(rows=(-1,))  # note the single element tuple here

# SELECT AVG(col) OVER (ROWS BETWEEN 1 PRECEDING AND CURRENT ROW) FROM t
t.col.mean().over(rows=-1)  # note the integer here

# SELECT AVG(col) OVER (ORDER BY time RANGE BETWEEN INTERVAL '7' DAYS PRECEDING AND CURRENT ROW)
t.col.mean().over(range=-interval(days=7), order_by="time")

# SELECT AVG(col) OVER (PARTITION BY a)
t.col.mean().over(group_by="time")

# SELECT AVG(col) OVER (ROWS BETWEEN 10 PRECEDING AND 3 FOLLOWING)
t.col.mean().over(rows=(-10, 3))

# SELECT AVG(col) OVER (ROWS BETWEEN 5 PRECEDING AND 3 PRECEDING)
t.col.mean().over(rows=(-5, -3))
```

The convenience functions `trailing_window` and `cumulative_window` would be turned into methods on scalar expressions called `trailing` and `cumulative` and those methods would take the corresponding arguments that their current counterparts take.

This would be an API breaking change, and would therefore need to go into a major release.

I think this simplifies the API a bit, but it might not be so much that it's worth doing.

### LINKED ISSUE #3305: bug(api): windows do not accept expressions for bounds
Right now the following lines of code do not work:

```python
import ibis

ibis.range_window(following=(ibis.interval(seconds=1), None))
ibis.window(following=(ibis.literal(1), None))
```

with this traceback:

```
Traceback (most recent call last):
  File "/home/cloud/src/ibis/timeseries.py", line 3, in <module>
    ibis.range_window(following=(ibis.interval(seconds=1), None))
  File "/home/cloud/src/ibis/ibis/expr/window.py", line 393, in range_window
    return Window(
  File "/home/cloud/src/ibis/ibis/expr/window.py", line 134, in __init__
    self._validate_frame()
  File "/home/cloud/src/ibis/ibis/expr/window.py", line 188, in _validate_frame
    if start < 0:
  File "/home/cloud/src/ibis/ibis/expr/types.py", line 54, in __bool__
    raise ValueError(
ValueError: The truth value of an Ibis expression is not defined
```

We have an assumption that the bounds of a window are constants, and that is an incorrect assumption. We need to allow arbitrary expressions in and let the individual backends reject invalid code. There's no feasible, maintainable static analysis we can do here given the need to allow expressions in window bounds.

### PR DESCRIPTION
**Pull request by kszucs**
resolves #3305
mainly resolves #1834 we just need to expose the same arguments to `.over()`

possible follow-ups: 
 - add support for `GROUPS` frame
 - remove max lookback support
 - deprecate legacy API
 - support for deferred objects (a more generic approach would be nice)
 - update the documentation

### CODE REVIEW

File: ibis/tests/expr/test_window_functions.py [hunk 1]
```diff
@@ -344,4 +12,19 @@ def test_quantile_shape():
     expr = t.projection(projs)
     (b1,) = expr.op().selections
 
-    assert b1.output_shape == rlz.Shape.COLUMNAR
+    assert b1.output_shape.is_columnar()
+
+
+# FIXME(kszucs)
```
Comment (kszucs) [inline ibis/tests/expr/test_window_functions.py:18]: Need to support deferred objects in the window API.

File: ibis/tests/expr/test_window_functions.py [hunk 2]
```diff
@@ -344,4 +12,19 @@ def test_quantile_shape():
...
+# def test_something(alltypes):
+#     t = alltypes
+#     t.group_by(t.g).mutate(ibis.row_number().name('foo'))
+
+
+# TODO(kszucs): add this test case without the rendered impala sql
```
Comment (kszucs) [inline ibis/tests/expr/test_window_functions.py:24]: It has been already added to test_window_frames.py, remove it from here.

File: ibis/backends/pandas/execution/window.py [hunk 3]
```diff
@@ -169,13 +165,13 @@ def get_aggcontext_window(
...
-            max_lookback = window.max_lookback
+        if frame.start is not None:
             assert not isinstance(operand, ops.CumulativeOp)
             aggcontext = agg_ctx.Moving(
-                preceding,
-                max_lookback,
+                frame.start,
+                # FIXME(kszucs): I don't think that we have a proper max_lookback test
+                # case because passing None here is not braking anything
+                frame.max_lookback,
```
Comment (kszucs) [inline ibis/backends/pandas/execution/window.py:174]: Could use a test case here.

File: ibis/backends/impala/tests/test_window.py [hunk 4]
```diff
@@ -146,14 +129,15 @@ def test_order_by_desc(alltypes, snapshot):
     assert_sql_equal(expr, snapshot, "out2.sql")
 
 
-def test_row_number_does_not_require_order_by(alltypes, snapshot):
-    t = alltypes
+# FIXME(kszucs)
+# def test_row_number_does_not_require_order_by(alltypes, snapshot):
```
Comment (kszucs) [inline ibis/backends/impala/tests/test_window.py:133]: Re-enable this test case.

File: ibis/backends/impala/tests/test_exprs.py [hunk 5]
```diff
@@ -508,41 +508,42 @@ def test_aggregations(alltypes):
...
-        f.mean(),
-        f.min(),
-        f.max(),
-    ]
-
-    proj_exprs = [expr.name('e%d' % i) for i, expr in enumerate(exprs)]
-
-    proj_table = g.mutate(proj_exprs)
-    proj_table.execute()
+# FIXME(kszucs)
```
Comment (kszucs) [inline ibis/backends/impala/tests/test_exprs.py:511]: Re-enable this test case. Also have a backendless version among the expression tests.

File: ibis/tests/expr/test_window_frames.py [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+    assert w7.end == ops.WindowBoundary(value, preceding=False)
+    assert w7.how == 'range'
+
+    w8 = w0.between(5, ibis.interval(seconds=11))
+    value = ibis.literal(5).cast("interval<int8>('s')")
+    assert w8.start == ops.WindowBoundary(value, preceding=False)
+    assert w8.end == ops.WindowBoundary(ibis.interval(seconds=11), preceding=False)
+    assert w8.how == 'range'
+
+    w9 = w0.between(-0.5, 1.5)
```
Comment (cpcloud) [inline ibis/tests/expr/test_window_frames.py:170]: Omg this is much better :)

File: ibis/tests/expr/test_window_frames.py [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+#     assert w3 == expected
+
+#     w4 = ibis.range_window(following=ibis.interval(days=5), order_by=t.e)
+#     w5 = ibis.range_window(following=ibis.interval(days=7), order_by=t.f)
+#     expected = ibis.range_window(following=ibis.interval(days=5), order_by=[t.e, t.f])
+#     w6 = w4.combine(w5)
+#     assert w6 == expected
+
+
+# def test_combine_windows(alltypes):
```
Comment (cpcloud) [inline ibis/tests/expr/test_window_frames.py:462]: Should any of these tests be uncommented?

File: ibis/tests/expr/test_window_frames.py [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+        param(dict(following=(2, -1)), id="invalid_following"),
+    ],
+)
+def test_window_api_preceding_following_invalid(case):
+    with pytest.raises(IbisInputError):
+        ibis.window(**case)
+
+    # FIXME(kszucs)
+    # with pytest.raises(IbisInputError):
+    #     ibis.range_window(**case)
```
Comment (cpcloud) [inline ibis/tests/expr/test_window_frames.py:304]: Should this be uncommented?

**Review by cpcloud — COMMENTED** [review]
Great work!

File: ibis/backends/base/sql/alchemy/registry.py [hunk 9]
```diff
@@ -303,69 +302,73 @@ def _endswith(t, op):
...
-    return an.windowize_function(new_expr, win)
+    return an.windowize_function(new_expr, frame=new_frame)
+
+
+def _translate_window_boundary(boundary):
+    if boundary is None:
+        return None
 
+    if isinstance(boundary.value, ops.Literal):
+        if boundary.preceding:
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/registry.py:323]: `boundary.preceding` is always a python bool.

`boundary.value` is either `None` or an `ops.Value`

File: ibis/backends/impala/tests/test_exprs.py (hunk 5 above)
Comment (kszucs) [reply ibis/backends/impala/tests/test_exprs.py]: Done.

File: ibis/backends/impala/tests/test_window.py (hunk 4 above)
Comment (kszucs) [reply ibis/backends/impala/tests/test_window.py]: Done.

File: ibis/backends/impala/tests/test_analytic_functions.py [hunk 10]
```diff
@@ -12,10 +11,11 @@ def table(mockcon):
 @pytest.mark.parametrize(
     "expr_fn",
     [
-        pytest.param(
-            lambda t: ibis.row_number().over(ibis.window(order_by=t.float_col)),
-            id="row_number",
-        ),
+        # FIXME(kszucs)
+        # pytest.param(
+        #     lambda t: ibis.row_number().over(ibis.window(order_by=t.float_col)),
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/impala/tests/test_analytic_functions.py]: Actually I removed this test case because it returns with a `Deferred` since it's unbound.

File: ibis/backends/bigquery/tests/unit/test_compiler.py [hunk 11]
```diff
@@ -453,12 +453,12 @@ def test_window_function(alltypes, window, snapshot):
...
-                preceding=(4, 2), group_by="year", order_by="timestamp_col"
-            ),
-            id="prec_prec",
-        ),
+        # param(
+        #     ibis.range_window(
+        #         preceding=(4, 2), group_by="year", order_by="timestamp_col"
+        #     ),
+        #     id="prec_prec",
+        # ),
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/bigquery/tests/unit/test_compiler.py]: Uncommented.

File: ibis/tests/expr/test_window_frames.py (hunk 8 above)
Comment (kszucs) [reply ibis/tests/expr/test_window_frames.py:326]: Uncommented and fixed.

File: ibis/expr/builders.py [hunk 12]
```diff
@@ -80,3 +86,114 @@ def when(self, case_expr, result_expr):
...
+    def group_by(self, expr):
+        return self.copy(groupings=self.groupings + util.promote_tuple(expr))
+
+    def order_by(self, expr):
+        return self.copy(orderings=self.orderings + util.promote_tuple(expr))
+
+    def lookback(self, value):
+        return self.copy(max_lookback=value)
+
+    # def combine(self, frame):
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/builders.py:185]: Yes, removed.

File: ibis/expr/api.py [hunk 13]
```diff
@@ -927,10 +932,293 @@ def get_backend(expr: Expr | None = None) -> BaseBackend:
...
+        if is_negative(following):
+            raise IbisInputError("following end point must be non-negative")
+        between = (None, _following(following))
+
+    w = bl.WindowBuilder().group_by(group_by).order_by(order_by).lookback(max_lookback)
+    if between is not None:
+        return w.between(*between)
+    elif rows is not None:
+        return w.rows(*rows)
+    elif range is not None:
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/api.py]: Now I enforced it.

File: ibis/backends/base/sql/alchemy/registry.py [hunk 14]
```diff
@@ -303,69 +302,73 @@ def _endswith(t, op):
...
 
-def _cumulative_to_window(translator, op, window):
-    win = W.cumulative_window()
-    win = win.group_by(window._group_by).order_by(window._order_by)
-
+def _cumulative_to_window(translator, op, frame):
     klass = _cumulative_to_reduction[type(op)]
     new_op = klass(*op.args)
     new_expr = new_op.to_expr().name(op.name)
+    new_frame = frame.copy(how='rows', start=None, end=0)
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/registry.py:306]: Yes, updated.

File: ibis/expr/operations/window.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    @property
+    def name(self) -> str:
+        self.value.name
+
+    @property
+    def following(self) -> bool:
+        return not self.preceding
+
+
+# perhaps have separate window frames, RowsWindowFrame and RangeWindowFrame
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/operations/window.py:32]: Separated them.

File: ibis/expr/types/generic.py [hunk 16]
```diff
@@ -482,6 +491,12 @@ def desc(self) -> ir.Value:
         """Sort an expression descending."""
         return ops.SortKey(self, ascending=False).to_expr()
 
+    def preceding(self) -> ir.Value:
+        return ops.WindowBoundary(self, preceding=True).to_expr()
+
+    def following(self) -> ir.Value:
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/types/generic.py]: Removed them. I added them to support deferred objects using the same approach with sortkeys, but it requires way more boilerplate for the window builder due to various validations.

We can add support for deferred objects in follow-ups.

File: ibis/tests/expr/test_window_frames.py (hunk 7 above)
Comment (kszucs) [reply ibis/tests/expr/test_window_frames.py:507]: We don't have the `combine()` method anymore, so I removed them.

File: ibis/backends/base/sql/alchemy/registry.py [hunk 17]
```diff
@@ -638,12 +641,13 @@ def translator(t, op: ops.Node):
...
     ops.CumeDist: unary(lambda _: sa.func.cume_dist()),
     ops.NthValue: _nth_value,
-    ops.Window: _window,
-    ops.CumulativeOp: _window,
+    ops.WindowFunction: _window_function,
     ops.CumulativeMax: unary(sa.func.max),
     ops.CumulativeMin: unary(sa.func.min),
     ops.CumulativeSum: unary(sa.func.sum),
     ops.CumulativeMean: unary(sa.func.avg),
+    # ops.CumulativeAny
```
Comment (kszucs) [inline ibis/backends/base/sql/alchemy/registry.py:649]: Add these

Comment (kszucs) [reply ibis/backends/base/sql/alchemy/registry.py]: Updated

File: ibis/expr/builders.py [hunk 18]
```diff
@@ -80,3 +86,114 @@ def when(self, case_expr, result_expr):
...
+    This class is patterned after SQL window frame clauses.
+
+    Using `None` for `preceding` or `following` indicates an unbounded frame.
+
+    Use 0 for `CURRENT ROW`.
+    """
+
+    __type__ = ops.WindowFrame
+
+    how = rlz.optional(rlz.isin({'rows', 'range'}), default="rows")
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/expr/builders.py:99]: Kept the rows default as discussed.

File: ibis/backends/base/sql/alchemy/registry.py [hunk 19]
```diff
@@ -300,69 +299,73 @@ def _endswith(t, op):
...
+                'backends.'
+            )
+        how = 'rows'
+    elif isinstance(window.frame, ops.RangeWindowFrame):
+        how = 'range_'
+    else:
+        raise NotImplementedError(type(window.frame))
+
+    if t._forbids_frame_clause and isinstance(window.func, t._forbids_frame_clause):
+        # some functions don't support frame clauses
```
Comment (cpcloud) [inline ibis/backends/base/sql/alchemy/registry.py:357]: It's more accurate to say "some functions on some backends". No need to change anything, just want to capture that it's both a limitation/design choice by the backend as well as the particular function.

File: ibis/backends/base/sql/registry/window.py [hunk 20]
```diff
@@ -1,34 +1,19 @@
...
-
-import ibis
 import ibis.common.exceptions as com
 import ibis.expr.analysis as an
 import ibis.expr.datatypes as dt
 import ibis.expr.operations as ops
-import ibis.expr.types as ir
 
 _map_interval_to_microseconds = {
+    'Y': 31536000000000,
```
Comment (cpcloud) [inline ibis/backends/base/sql/registry/window.py:9]: Do we have a use case for this yet?

File: ibis/backends/base/sql/registry/window.py [hunk 21]
```diff
@@ -42,201 +27,140 @@
...
+
+    frame = frame.copy(start=None, end=0)
+    expr = an.windowize_function(func.to_expr(), frame)
+    return expr.op()
 
 
-def cumulative_to_window(translator, op, window):
-    klass = _cumulative_to_reduction[type(op)]
-    new_op = klass(*op.args)
+def interval_boundary_to_integer(boundary):
```
Comment (cpcloud) [inline ibis/backends/base/sql/registry/window.py:46]: Looks like this function is missing some test coverage. Can we delete the uncovered code or add a test?

File: ibis/backends/base/sql/registry/window.py [hunk 22]
```diff
@@ -42,201 +27,140 @@
...
 
     # Some analytic functions need to have the expression of interest in
     # the ORDER BY part of the window clause
-    if isinstance(arg, translator._require_order_by) and not window._order_by:
-        window = window.order_by(arg.args[0])
+    frame = op.frame
+    if isinstance(op.func, translator._require_order_by) and not frame.order_by:
+        frame = frame.copy(order_by=(op.func.arg,))
 
     # Time ranges need to be converted to microseconds.
```
Comment (cpcloud) [inline ibis/backends/base/sql/registry/window.py:147]: Is this comment still accurate? If you haven't already done it in this PR we should remove the microseconds assumption here and try to move that logic into the specific backend that needs it (I believe it's BigQuery?).

File: ibis/backends/bigquery/tests/unit/snapshots/test_compiler/test_trailing_range_window/days/out.sql [hunk 23]
```diff
@@ -1,3 +1,3 @@
 SELECT t0.*,
-       avg(t0.`float_col`) OVER (ORDER BY UNIX_MICROS(t0.`timestamp_col`) RANGE BETWEEN 86400000000 PRECEDING AND CURRENT ROW) AS `win_avg`
+       avg(t0.`float_col`) OVER (ORDER BY UNIX_MICROS(t0.`timestamp_col`) ASC RANGE BETWEEN 86400000000 PRECEDING AND EXTRACT(DAY from INTERVAL 0 DAY) * 86400000000 FOLLOWING) AS `win_avg`
```
Comment (cpcloud) [inline ibis/backends/bigquery/tests/unit/snapshots/test_compiler/test_trailing_range_window/days/out.sql:2]: This is pretty gnarly. Is the `EXTRACT` bit actually equivalent to `CURRENT ROW` 😅 ?

Comment (cpcloud) [reply ibis/backends/bigquery/tests/unit/snapshots/test_compiler/test_trailing_range_window/days/out.sql:2]: I guess so, since `0 FOLLOWING = CURRENT ROW`!

File: ibis/backends/bigquery/tests/unit/test_compiler.py [hunk 24]
```diff
@@ -470,14 +470,15 @@ def test_range_window_function(alltypes, window, snapshot):
     "preceding",
     [
         param(5, id="five"),
-        param(ibis.interval(nanoseconds=1), id="nanos"),
+        param(ibis.interval(nanoseconds=1), id="nanos", marks=pytest.mark.xfail),
```
Comment (cpcloud) [inline ibis/backends/bigquery/tests/unit/test_compiler.py:473]: Was this ever correctly passing?

File: ibis/backends/clickhouse/compiler/values.py [hunk 25]
```diff
@@ -1173,32 +1155,21 @@ def _foll(f: int | None) -> str:
...
-            kind = 'ROWS' if p > 0 else 'RANGE'
-            frame = f'{kind} BETWEEN {_prec(p)} AND UNBOUNDED FOLLOWING'
-    elif f is not None:
-        if isinstance(f, tuple):
-            start, end = map(_foll, f)
-            frame = f'{window.how.upper()} BETWEEN {start} AND {end}'
+            start = format_window_boundary(frame.start, **kw)
+
+        if frame.end is None:
+            end = 'UNBOUNDED FOLLOWING'
```
Comment (cpcloud) [inline ibis/backends/clickhouse/compiler/values.py:1168]: Do we need a test for this?

File: ibis/backends/impala/tests/test_analytic_functions.py [hunk 26]
```diff
@@ -12,10 +11,6 @@ def table(mockcon):
 @pytest.mark.parametrize(
     "expr_fn",
     [
-        pytest.param(
-            lambda t: ibis.row_number().over(ibis.window(order_by=t.float_col)),
-            id="row_number",
-        ),
```
Comment (cpcloud) [inline ibis/backends/impala/tests/test_analytic_functions.py:18]: Is this test bogus in some way? Why did you remove it?

File: ibis/backends/pandas/execution/window.py [hunk 27]
```diff
@@ -246,25 +247,31 @@ def trim_window_result(data: pd.Series | pd.DataFrame, timecontext: TimeContext
     return indexed_subset[name]
 
 
-@execute_node.register(ops.Window, pd.Series, win.Window)
+@execute_node.register(ops.WindowBoundary, int, bool)
+def execute_window_boundary_int(op, data, _, **kwargs):
+    return data
```
Comment (cpcloud) [inline ibis/backends/pandas/execution/window.py:252]: Do we need this dispatch?

File: ibis/expr/api.py [hunk 28]
```diff
@@ -507,7 +503,15 @@ def asc(expr: ir.Column | str) -> ir.Value:
...
     """
-    return _sort_order(expr, "asc")
+    return _deferred_method_call(expr, "asc")
+
+
+def preceding(value) -> ir.Value:
+    return ops.WindowBoundary(value, preceding=True).to_expr()
+
+
+def following(value) -> ir.Value:
```
Comment (cpcloud) [inline ibis/expr/api.py:513]: I thought the plan was to make these private to avoid confusion until we know what that boundary operations do when someone calls `.execute()` on them (if anything).

File: ibis/expr/builders.py [hunk 29]
```diff
@@ -80,3 +82,188 @@ def when(self, case_expr, result_expr):
...
+class LegacyWindowBuilder(WindowBuilder):
+    def _is_negative(self, value):
+        if value is None:
+            return False
+        if isinstance(value, ir.Scalar):
+            value = value.op().value
+        return value < 0
+
+    def preceding_following(self, preceding, following, how=None):
+        preceding_tuple = has_preceding = False
```
Comment (cpcloud) [inline ibis/expr/builders.py:211]: Whew, this one's a doozy!

File: ibis/expr/operations/window.py [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+    # pretty similar to SortKey and Alias operations which wrap a single value
+    value = rlz.one_of([rlz.numeric, rlz.interval])
+    preceding = rlz.bool_
+
+    output_shape = rlz.shape_like("value")
+    output_dtype = rlz.dtype_like("value")
+
+    @property
+    def name(self) -> str:
+        self.value.name
```
Comment (cpcloud) [inline ibis/expr/operations/window.py:25]:
```suggestion
        return self.value.name
```

and this probably needs a test to catch things like a missing `return` :)

File: ibis/tests/expr/test_value_exprs.py [hunk 31]
```diff
@@ -1662,3 +1641,16 @@ def test_where_output_shape():
...
+
+
+def test_quantile_shape():
+    t = ibis.table([("a", "float64")])
+
+    b1 = t.a.quantile(0.25).name("br2")
+    assert isinstance(b1, ir.Scalar)
+
+    projs = [b1]
+    expr = t.projection(projs)
```
Comment (cpcloud) [inline ibis/tests/expr/test_value_exprs.py:1653]:
```suggestion
    expr = t.select(projs)
```
Try to avoid using `projection`, we should move away from it as the conventional API to select/compute columns.

**Review by cpcloud — CHANGES_REQUESTED** [review]
A remaining questions/comments! This is ready to go after this round of review.

File: ibis/backends/base/sql/alchemy/registry.py (hunk 19 above)
Comment (kszucs) [reply ibis/backends/base/sql/alchemy/registry.py:357]: Updated the comment.

File: ibis/backends/base/sql/registry/window.py (hunk 20 above)
Comment (kszucs) [reply ibis/backends/base/sql/registry/window.py:9]: Not really, going to remove it to avoid issues with leap years.

Comment (kszucs) [reply ibis/backends/base/sql/registry/window.py:9]: Removed.

File: ibis/expr/builders.py (hunk 29 above)
Comment (kszucs) [reply ibis/expr/builders.py:211]: It is!

File: ibis/expr/operations/window.py (hunk 30 above)
Comment (kszucs) [reply ibis/expr/operations/window.py:25]: Nice catch! Actually we don't need to implement `name` here since `ops.Value` provides a default implementation.

File: ibis/tests/expr/test_value_exprs.py (hunk 31 above)
Comment (kszucs) [reply ibis/tests/expr/test_value_exprs.py:1653]: Updated.

File: ibis/backends/base/sql/registry/window.py (hunk 21 above)
Comment (kszucs) [reply ibis/backends/base/sql/registry/window.py:45]: Covered by the bigquery tests.

File: ibis/backends/base/sql/registry/window.py (hunk 22 above)
Comment (kszucs) [reply ibis/backends/base/sql/registry/window.py:146]: Yes, it is the bigquery backend. I'd rather do that in the follow-up, since I cannot run the impala tests locally which uses the same base backend.

Comment (kszucs) [reply ibis/backends/base/sql/registry/window.py:146]: Created follow-up https://github.com/ibis-project/ibis/issues/5419

File: ibis/backends/impala/tests/test_analytic_functions.py (hunk 26 above)
Comment (kszucs) [reply ibis/backends/impala/tests/test_analytic_functions.py:18]: `ibis.row_number()` now returns with a deferred since it's not bound to any table yet

calling over returns with a deferred as well since we are trying to bind the window frame if the analytic expression is bound, otherwise we just defer this binding until we pass the expression to a projection

File: ibis/backends/pandas/execution/window.py (hunk 27 above)
Comment (kszucs) [reply ibis/backends/pandas/execution/window.py:252]: No, we don't, removing. Nice catch!

File: ibis/expr/api.py (hunk 28 above)
Comment (kszucs) [reply ibis/expr/api.py:513]: This is a way to explicitly mark expressions as preceding or following if the inference in the `window_boundary` rule wouldn't apply, imagine the following scenario:

```python
start = -myexpr  # ops.Netaget(...something)

ibis.window(rows=(start, t.int_col + 10))
# ROWS BETWEEN <myexpr> PRECEDING AND t.int_col + 10 FOLLOWING`
```

we infer the negate operation as a preceding marker, though the user could have thought about `ROWS BETWEEN <-myexpr> FOLLOWING AND t.int_col + 10 FOLLOWING` in which case wrapping the start expr with `ibis.following()` would ensure that:

```python
ibis.window(rows=(ibis.following(start), t.int_col + 10))
# ROWS BETWEEN -myexpr FOLLOWING AND t.int_col + 10 FOLLOWING`
```

Comment (kszucs) [reply ibis/expr/api.py:513]: So `ibis.preceding()` and `ibis.following()` are for forcing the start/end boundaries instead of letting the window boundary rule to figure it out.

File: ibis/backends/bigquery/tests/unit/test_compiler.py (hunk 24 above)
Comment (kszucs) [reply ibis/backends/bigquery/tests/unit/test_compiler.py:473]: It successfully compiled but would fail to execute.

File: ibis/backends/clickhouse/compiler/values.py (hunk 25 above)
Comment (kszucs) [reply ibis/backends/clickhouse/compiler/values.py:1168]: Added  a simple test case.

**Comment by kszucs** [comment]
@cpcloud can you please take another look? Shall I squash this into a single commit?

**Review by cpcloud — APPROVED** [review]
Great work! 🚢 🪟

**Comment by cpcloud** [comment]
@kszucs Squash as you like. I would keep the things that are strictly new features or bug fixes if that's easy, otherwise yeah just squash it into a single commit.
